File: VerbumKit/Sources/Features/Context/ContextFeature.swift
import Clients
import ComposableArchitecture
import Models

@Reducer
public struct ContextFeature {
    @ObservableState
    public struct State: Equatable {
        public let reference: PassageReference
        public var content: Content = .idle

        public init(reference: PassageReference) {
            self.reference = PassageReference(bookId: reference.bookId, chapter: reference.chapter)
        }
    }

    public enum Content: Equatable, Sendable {
        case idle, loading, unavailable, failed
        case loaded(PassageContext)
    }

    public enum Action: Equatable {
        case task, retryTapped
        case response(Result<PassageContext?, Failure>)
        case passageTapped(PassageReference)
        case entityTapped(BibleEntity)
        case delegate(Delegate)

        @CasePathable
        public enum Delegate: Equatable {
            case openPassage(PassageReference)
            case openEntity(BibleEntity)
        }
    }

    public struct Failure: Error, Equatable, Sendable {}
    @Dependency(\.contextClient) var contextClient
    public init() {}

    public var body: some ReducerOf<Self> {
        Reduce { state, action in
            switch action {
            case .task:
                guard state.content == .idle else { return .none }
                return load(&state)
            case .retryTapped:
                return load(&state)
            case .response(.success(let context)):
                state.content = context.map(Content.loaded) ?? .unavailable
                return .none
            case .response(.failure):
                state.content = .failed
                return .none
            case .passageTapped(let reference):
                return .send(.delegate(.openPassage(reference)))
            case .entityTapped(let entity):
                return .send(.delegate(.openEntity(entity)))
            case .delegate:
                return .none
            }
        }
    }

    private func load(_ state: inout State) -> Effect<Action> {
        state.content = .loading
        return .run { [contextClient, reference = state.reference] send in
            do {
                let context = try await contextClient.chapter(reference)
                try Task.checkCancellation()
                await send(.response(.success(context)))
            } catch is CancellationError {
                return
            } catch {
                await send(.response(.failure(Failure())))
            }
        }
        .cancellable(id: CancelID.load, cancelInFlight: true)
    }

    private enum CancelID { case load }
}

File: android/feature/scripture/src/main/kotlin/com/nexussoft/verbum/feature/scripture/ContextFeature.kt
package com.nexussoft.verbum.feature.scripture

import com.nexussoft.verbum.clients.ContextClient
import com.nexussoft.verbum.common.arch.*
import com.nexussoft.verbum.models.*
import kotlinx.coroutines.CancellationException

object ContextFeature {
    data class State(val reference: PassageReference, val content: Content = Content.Idle)
    sealed interface Content {
        data object Idle : Content
        data object Loading : Content
        data object Unavailable : Content
        data object Failed : Content
        data class Loaded(val page: PassageContext) : Content
    }
    sealed interface Action {
        data object Started : Action
        data object RetryTapped : Action
        data class Response(val page: PassageContext?) : Action
        data object Failed : Action
        data class PassageTapped(val reference: PassageReference) : Action
        data class EntityTapped(val entity: BibleEntity) : Action
        data class Delegate(val delegate: DelegateAction) : Action
    }
    sealed interface DelegateAction {
        data class OpenPassage(val reference: PassageReference) : DelegateAction
        data class OpenEntity(val entity: BibleEntity) : DelegateAction
    }

    fun reducer(client: ContextClient): Reducer<State, Action> = Reducer { state, action ->
        when (action) {
            Action.Started, Action.RetryTapped -> {
                if (action == Action.Started && state.content != Content.Idle) state.only()
                else state.copy(content = Content.Loading).with(runEffect { send ->
                    try {
                        send(Action.Response(client.chapter(state.reference)))
                    } catch (e: CancellationException) {
                        throw e
                    } catch (_: Exception) {
                        send(Action.Failed)
                    }
                })
            }
            is Action.Response -> state.copy(content = action.page?.let { Content.Loaded(it) } ?: Content.Unavailable).only()
            Action.Failed -> state.copy(content = Content.Failed).only()
            is Action.PassageTapped -> state.with(Effect.Send(Action.Delegate(DelegateAction.OpenPassage(action.reference))))
            is Action.EntityTapped -> state.with(Effect.Send(Action.Delegate(DelegateAction.OpenEntity(action.entity))))
            is Action.Delegate -> state.only()
        }
    }
}
